### PR TITLE
Initial commit for Idea Hub draft ideas and published ideas selectors.

### PR DESCRIPTION
**Pull request by asvinb**
## Summary

Addresses issue #3270 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/idea-hub/datastore/__fixtures__/draft-post-ideas.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
+[
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/__fixtures__/draft-post-ideas.json:1]: Let's update `draft-post-ideas.json` and `published-post-ideas.json` files to use the correct format for its items which must include `postID`, `postEditURL`, and `postURL` (for published posts only) fields.

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	reducerCallback: ( state, draftPostIdeas ) => {
+		return {
+			...state,
+			draftPostIdeas,
+		};
+	},
+} );
+
+const baseInitialState = {
+	draftPostIdeas: [],
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/draft-post-ideas.js:41]: The initial value should be `undefined` which indicates that data hasn't been loaded yet.

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	*getDraftPostIdeas( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const draftPostIdeas = registry.select( STORE_NAME ).getDraftPostIdeas( options );
+
+		// If there are already draft ideas in state, don't make an API request.
+		if ( draftPostIdeas.length ) {
+			return;
+		}
+
+		yield fetchGetDraftPostIdeasStore.actions.fetchGetDraftPostIdeas();
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/draft-post-ideas.js:54]: We need to check that `draftPostIdeas` doesn't equal `undefined`.

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	 * @param {number} [options.offset] Optional. From which array index to get ideas.
+	 * @param {number} [options.length] Optional. Amount of draft post ideas to return.
+	 * @return {(Array.<Object>|undefined)} A list of idea hub ideas; `undefined` if not loaded.
+	 */
+	getDraftPostIdeas( state, options = {} ) {
+		const { draftPostIdeas } = state;
+		const offset = options?.offset || 0;
+		const length = options.length ? offset + options.length : draftPostIdeas.length;
+		return ( 'offset' in options || 'length' in options ) ? draftPostIdeas.slice( offset, length ) : draftPostIdeas;
+	},
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/draft-post-ideas.js:75]: It should return `undefined` if ideas aren't loaded yet.

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	reducerCallback: ( state, publishedPostIdeas ) => {
+		return {
+			...state,
+			publishedPostIdeas,
+		};
+	},
+} );
+
+const baseInitialState = {
+	publishedPostIdeas: [],
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/published-post-ideas.js:41]: Same here, it should be `undefined` by default.

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	*getPublishedPostIdeas( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const publishedPostIdeas = registry.select( STORE_NAME ).getPublishedPostIdeas( options );
+
+		// If there are already published ideas in state, don't make an API request.
+		if ( publishedPostIdeas.length ) {
+			return;
+		}
+
+		yield fetchGetPublishedPostIdeasStore.actions.fetchGetPublishedPostIdeas();
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/published-post-ideas.js:54]: And here, we need to make a new request only if `publishedPostIdeas` equals `undefined`.

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	 * @param {number} [options.offset] Optional. From which array index to get ideas.
+	 * @param {number} [options.length] Optional. Amount of published post ideas to return.
+	 * @return {(Array.<Object>|undefined)} A list of idea hub ideas; `undefined` if not loaded.
+	 */
+	getPublishedPostIdeas( state, options = {} ) {
+		const { publishedPostIdeas } = state;
+		const offset = options?.offset || 0;
+		const length = options.length ? offset + options.length : publishedPostIdeas.length;
+		return ( 'offset' in options || 'length' in options ) ? publishedPostIdeas.slice( offset, length ) : publishedPostIdeas;
+	},
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/published-post-ideas.js:75]: It should return `undefined` if `publishedPostIdeas` are not loaded yet.

File: includes/Modules/Idea_Hub.php [hunk 8]
```diff
@@ -277,6 +316,50 @@ protected function create_data_request( Data_Request $data ) {
...
+					function( $post_id ) {
+						return array_merge(
+							array(
+								'postID'      => $post_id,
+								'postEditURL' => get_edit_post_link( $post_id ),
+							),
+							$this->get_post_idea( $post_id )
+						);
+					},
+					$response
```
Comment (eugene-manuilov) [inline includes/Modules/Idea_Hub.php:342]: Let's make sure this is an array and convert it to an array if it is not:
```suggestion
					is_array( $response ) ? $response : array( $response )
```

File: includes/Modules/Idea_Hub.php [hunk 9]
```diff
@@ -277,6 +316,50 @@ protected function create_data_request( Data_Request $data ) {
...
+						return array_merge(
+							array(
+								'postID'      => $post_id,
+								'postEditURL' => get_edit_post_link( $post_id ),
+								'postURL'     => get_permalink( $post_id ),
+							),
+							$this->get_post_idea( $post_id )
+						);
+					},
+					$response
```
Comment (eugene-manuilov) [inline includes/Modules/Idea_Hub.php:356]: Same here, let's make sure it is an array:
```suggestion
					is_array( $response ) ? $response : array( $response )
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @asvinb. It is a very good start. Added a few comments with some tweaks that are required before we can approve this PR. Could you please take a look at it and make the appropriate changes? Plus, update js tests to reflect changes in selectors.

File: assets/js/modules/idea-hub/datastore/__fixtures__/draft-post-ideas.json (hunk 1 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/__fixtures__/draft-post-ideas.json:1]: Updated @eugene-manuilov

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js (hunk 2 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/draft-post-ideas.js]: Updated @eugene-manuilov Can you check?

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js (hunk 3 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/draft-post-ideas.js]: Updated @eugene-manuilov Can you check?

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js (hunk 4 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/draft-post-ideas.js]: Updated @eugene-manuilov Can you check?

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js (hunk 7 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/published-post-ideas.js]: Updated @eugene-manuilov Can you check?

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js (hunk 6 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/published-post-ideas.js]: Updated @eugene-manuilov Can you check?

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js (hunk 5 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/published-post-ideas.js]: Updated @eugene-manuilov Can you check?

File: includes/Modules/Idea_Hub.php (hunk 8 above)
Comment (asvinb) [reply includes/Modules/Idea_Hub.php]: Updated @eugene-manuilov Can you check?

File: includes/Modules/Idea_Hub.php (hunk 9 above)
Comment (asvinb) [reply includes/Modules/Idea_Hub.php]: Updated @eugene-manuilov Can you check?

**Comment by asvinb** [comment]
@eugene-manuilov I've updated the PR following your feedback. About setting the data to `undefined`, I think we should make those changes to the `saved-ideas` and `new-ideas` data store partial as well, because they all work the same way.

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	*getDraftPostIdeas( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const draftPostIdeas = registry.select( STORE_NAME ).getDraftPostIdeas( options );
+
+		// If there are already draft ideas in state, don't make an API request.
+		if ( draftPostIdeas?.length ) {
+			return;
+		}
+
+		yield fetchGetDraftPostIdeasStore.actions.fetchGetDraftPostIdeas();
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/draft-post-ideas.js:54]: We should strictly check whether the `draftPostIdeas` is undefined to make sure we don't send unnecessary requests when `draftPostIdeas` is an empty array (which is a possible value).

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+	 * @param {Object} options          Options for getting draft post ideas.
+	 * @param {number} [options.offset] Optional. From which array index to get ideas.
+	 * @param {number} [options.length] Optional. Amount of draft post ideas to return.
+	 * @return {(Array.<Object>|undefined)} A list of idea hub ideas; `undefined` if not loaded.
+	 */
+	getDraftPostIdeas( state, options = {} ) {
+		const { draftPostIdeas } = state;
+
+		if ( draftPostIdeas === undefined ) {
+			return;
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/draft-post-ideas.js:74]: Let's explicitly return `undefined` to make it consistent with other selectors:
```suggestion
			return undefined;
```

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	*getPublishedPostIdeas( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const publishedPostIdeas = registry.select( STORE_NAME ).getPublishedPostIdeas( options );
+
+		// If there are already published ideas in state, don't make an API request.
+		if ( publishedPostIdeas?.length ) {
+			return;
+		}
+
+		yield fetchGetPublishedPostIdeasStore.actions.fetchGetPublishedPostIdeas();
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/published-post-ideas.js:54]: Same here we should send a request only if the `publishedPostIdeas` is undefined.

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+	 * @param {Object} options          Options for getting published post ideas.
+	 * @param {number} [options.offset] Optional. From which array index to get ideas.
+	 * @param {number} [options.length] Optional. Amount of published post ideas to return.
+	 * @return {(Array.<Object>|undefined)} A list of idea hub ideas; `undefined` if not loaded.
+	 */
+	getPublishedPostIdeas( state, options = {} ) {
+		const { publishedPostIdeas } = state;
+
+		if ( publishedPostIdeas === undefined ) {
+			return;
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/published-post-ideas.js:74]: Let's return `undefined` here as well:
```suggestion
			return undefined;
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @asvinb. Looks solid. Added a few more comments.

> ... I think we should make those changes to the saved-ideas and new-ideas data store partial as well, because they all work the same way.

Yes, let's update it there too.

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js (hunk 10 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/draft-post-ideas.js]: Can you check again @eugene-manuilov ?

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js (hunk 11 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/draft-post-ideas.js]: 👍

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js (hunk 12 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/published-post-ideas.js]: Can you check again @eugene-manuilov ?

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js (hunk 13 above)
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/published-post-ideas.js]: 👍

**Comment by asvinb** [comment]
@eugene-manuilov PR updated. Can you check?

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
All looks good apart from resolvers. We need to send API requests **only** when state values are not loaded yet (when they equal `undefined`). Could you please update resolvers and adjust js tests?

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.js [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+	*getDraftPostIdeas( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const draftPostIdeas = registry.select( STORE_NAME ).getDraftPostIdeas( options );
+
+		// If there are already draft ideas in state, don't make an API request.
+		if ( draftPostIdeas?.length || draftPostIdeas !== undefined ) {
+			return;
+		}
+
+		yield fetchGetDraftPostIdeasStore.actions.fetchGetDraftPostIdeas();
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/draft-post-ideas.js]: Ah I see what you mean. Updated @eugene-manuilov

File: assets/js/modules/idea-hub/datastore/new-ideas.js [hunk 15]
```diff
@@ -62,7 +47,7 @@ const baseResolvers = {
 		const newIdeas = registry.select( STORE_NAME ).getNewIdeas( options );
 
 		// If there are already new ideas in state, don't make an API request.
-		if ( newIdeas.length ) {
+		if ( newIdeas?.length || newIdeas !== undefined ) {
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/new-ideas.js]: 👍

File: assets/js/modules/idea-hub/datastore/published-post-ideas.js [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+	*getPublishedPostIdeas( options = {} ) {
+		const registry = yield Data.commonActions.getRegistry();
+		const publishedPostIdeas = registry.select( STORE_NAME ).getPublishedPostIdeas( options );
+
+		// If there are already published ideas in state, don't make an API request.
+		if ( publishedPostIdeas?.length || publishedPostIdeas !== undefined ) {
+			return;
+		}
+
+		yield fetchGetPublishedPostIdeasStore.actions.fetchGetPublishedPostIdeas();
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/published-post-ideas.js]: 👍

File: assets/js/modules/idea-hub/datastore/saved-ideas.js [hunk 17]
```diff
@@ -62,7 +47,7 @@ const baseResolvers = {
 		const savedIdeas = registry.select( STORE_NAME ).getSavedIdeas( options );
 
 		// If there are already saved ideas in state, don't make an API request.
-		if ( savedIdeas.length ) {
+		if ( savedIdeas?.length || savedIdeas !== undefined ) {
```
[the remark this answers is not in the document]
Comment (asvinb) [reply assets/js/modules/idea-hub/datastore/saved-ideas.js]: 👍

File: assets/js/modules/idea-hub/datastore/draft-post-ideas.test.js [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+
+			it( 'uses a resolver to make a network request', async () => {
+				fetchMock.getOnce(
+					/^\/google-site-kit\/v1\/modules\/idea-hub\/data\/draft-post-ideas/,
+					{ body: fixtures.draftPostIdeas, status: 200 }
+				);
+
+				const pendingDraftPostIdeas = registry.select( STORE_NAME ).getDraftPostIdeas( options );
+
+				expect( pendingDraftPostIdeas ).toEqual( undefined );
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/draft-post-ideas.test.js:67]:
```suggestion
				expect( pendingDraftPostIdeas ).toBeUndefined();
```

File: assets/js/modules/idea-hub/datastore/published-post-ideas.test.js [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+			it( 'uses a resolver to make a network request', async () => {
+				fetchMock.getOnce(
+					/^\/google-site-kit\/v1\/modules\/idea-hub\/data\/published-post-ideas/,
+					{ body: fixtures.publishedPostIdeas, status: 200 }
+				);
+
+				const pendingPublishedPostIdeas = registry.select( STORE_NAME ).getPublishedPostIdeas( options );
+
+				expect( pendingPublishedPostIdeas ).toEqual( undefined );
```
Comment (eugene-manuilov) [inline assets/js/modules/idea-hub/datastore/published-post-ideas.test.js:67]:
```suggestion
				expect( pendingPublishedPostIdeas ).toBeUndefined();
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Awesome! Looks good to me. Found two more minor issues. I'll apply it and approve this PR. Thanks, @asvinb.

**Pull request merged**